### PR TITLE
Replace sGet, sSave macros - part 3

### PR DESCRIPTION
**Pull request by dreamer**
Continuation of #550, #558, #564.

### CODE REVIEW

File: include/dos_inc.h [hunk 1]
```diff
@@ -645,25 +629,32 @@ class DOS_FCB : public MemStruct {
...
-	void SetPSPSeg(Bit16u _pspseg) { sSave(sMCB,psp_segment,_pspseg);}
-	Bit8u GetType(void) { return (Bit8u)sGet(sMCB,type);}
-	Bit16u GetSize(void) { return (Bit16u)sGet(sMCB,size);}
-	Bit16u GetPSPSeg(void) { return (Bit16u)sGet(sMCB,psp_segment);}
+
+	void SetType(uint8_t type) { SSET_BYTE(sMCB, type, type); }
+	uint8_t GetType() const { return SGET_BYTE(sMCB, type); }
+
+	void SetSize(uint16_t size) { SSET_WORD(sMCB, size, size); }
+	uint16_t GetSize() const { return SGET_WORD(sMCB, size); }
```
Comment (kcgen) [inline include/dos_inc.h:645]: Suggest a similar argument name change to differentiate the struct member from the argument in the "type, type" an "size, size" setter calls.

File: src/dos/dos_classes.cpp [hunk 2]
```diff
@@ -16,157 +19,113 @@
...
-}
-
-void DOS_InfoBlock::SetStartOfUMBChain(Bit16u _umbstartseg) {
-	sSave(sDIB,startOfUMBChain,_umbstartseg);
-}
-
-Bit8u DOS_InfoBlock::GetUMBChainState(void) {
-	return (Bit8u)sGet(sDIB,chainingUMB);
-}
-
```
Comment (kcgen) [inline src/dos/dos_classes.cpp:155]: (Assuming all of these are uneeded or were replaced somehow?)

File: include/dos_inc.h (hunk 1 above)
Comment (kcgen) [inline include/dos_inc.h:645]: Suggest similarly rename the arguments to differentiate them from the members. (ie: "size, size" and "type, type").

File: src/dos/dos_classes.cpp [hunk 3]
```diff
@@ -16,157 +19,113 @@
...
+	SSET_WORD(sDIB, sharingDelay, uint16_t(0));
+	SSET_WORD(sDIB, ptrCONinput, uint16_t(0)); // no unread input available
+	SSET_WORD(sDIB, maxSectorLength, uint16_t(0x200));
+	SSET_WORD(sDIB, dirtyDiskBuffers, uint16_t(0));
+	SSET_DWORD(sDIB, lookaheadBufPt, uint32_t(0));
+	SSET_WORD(sDIB, lookaheadBufNumber, uint16_t(0));
+	SSET_BYTE(sDIB, bufferLocation, uint8_t(0)); // buffer in base memory,
+	                                             // no workspace
+	SSET_DWORD(sDIB, workspaceBuffer, uint32_t(0));
+	SSET_WORD(sDIB, minMemForExec, uint16_t(0));
```
Comment (kcgen) [inline src/dos/dos_classes.cpp:95]: Looking at how almost all of these are fixed values, I wonder if we could keep a constexpr sDIB object pre-populated with all of these canned values:

constexpr sDIB set_location_defaults = { 
    regCXfrom5e = 0,
    countLRUcache = 0,
    countLRUopens = 0,
    ...
};
```

Then we assign it whenever `SetLocation` is called, and finish up with any dynamic values:

``` c++
void DOS_InfoBlock::SetLocation(uint16_t segment)
{
    ....
    sDIB obj = set_local_defaults;
    const RealPt dib_addr = RealMake(segment, offsetof(sDIB, diskBufferHeadPt));
    SSET_DWORD(sDIB, diskInfoBuffer, dib_addr);
    const uint16_t ext_size = static_cast<uint16_t>(MEM_TotalPages() * 4 - 1024);
    SSET_WORD(sDIB, extendedSize, ext_size);
```

File: include/dos_inc.h [hunk 4]
```diff
@@ -292,45 +289,23 @@ constexpr PhysPt assert_macro_args_ok()
...
-		case 2:mem_writew(pt+addr,(Bit16u)val);break;
-		case 4:mem_writed(pt+addr,(Bit32u)val);break;
-		}
-	}
-	void SetPt(Bit16u seg) { pt=PhysMake(seg,0);}
-	void SetPt(Bit16u seg,Bit16u off) { pt=PhysMake(seg,off);}
-	void SetPt(RealPt addr) { pt=Real2Phys(addr);}
+	MemStruct() = default;
+	MemStruct(uint16_t seg, uint16_t off) : pt(PhysMake(seg, off)) {}
+	MemStruct(RealPt addr) : pt(Real2Phys(addr)) {}
```
Comment (kcgen) [inline include/dos_inc.h:296]: Nice touches, as I see how these contribute to compile-time class initial values below!

File: src/dos/dos_classes.cpp [hunk 5]
```diff
@@ -423,11 +383,10 @@ void DOS_DTA::GetSearchParams(uint8_t &attr, char *pattern) const
...
 
 DOS_FCB::DOS_FCB(uint16_t seg, uint16_t off, bool allow_extended)
-        : extended(false),
-          real_pt(0)
+        : MemStruct(seg, off),
+          extended(false),
+          real_pt(pt)
 {
-	SetPt(seg, off);
-	real_pt = pt;
```
Comment (kcgen) [inline src/dos/dos_classes.cpp:430]: Very nice!

**Comment by kcgen** [comment]
Although out of exact scope of this PR, there are only a handful of implicit type-shortening warnings in `dos_classes.cpp` that could go into a separate commit.  Only concern is these in-theory could wrap; especially the ones that involve some arithmetic. Reported from clang v 10.x. 

``` c++
dos_classes.cpp:64:48: warning: implicit conversion loses integer precision: 'unsigned long' to 'PhysPt' (aka 'unsigned int') [-Wshorten-64-to-32]
        for(Bitu i=0;i<sizeof(sDIB);i++) mem_writeb(pt+i,0xff);
                                         ~~~~~~~~~~ ~~^~
dos_classes.cpp:65:38: warning: implicit conversion loses integer precision: 'unsigned long' to 'PhysPt' (aka 'unsigned int') [-Wshorten-64-to-32]
        for(Bitu i=0;i<14;i++) mem_writeb(pt+i,0);
                               ~~~~~~~~~~ ~~^~
In file included from dos_classes.cpp:22:
In file included from ../../include/dos_inc.h:22:
In file included from ../../include/dosbox.h:75:
../../include/logging.h:106:59: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
#define DEBUG_LOG_MSG(fmt, ...) fprintf(stderr, fmt "\n", ##__VA_ARGS__)
                                                          ^
dos_classes.cpp:333:55: warning: implicit conversion loses integer precision: 'unsigned long' to 'PhysPt' (aka 'unsigned int') [-Wshorten-64-to-32]
        for (i=0;i<11;i++) mem_writeb(pt+offsetof(sDTA,sname)+i,' ');
                           ~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~^~
dos_classes.cpp:403:41: warning: implicit conversion loses integer precision: 'unsigned long' to 'PhysPt' (aka 'unsigned int') [-Wshorten-64-to-32]
        for (i=0;i<fill;i++) mem_writeb(real_pt+i,0);
                             ~~~~~~~~~~ ~~~~~~~^~
dos_classes.cpp:503:24: warning: implicit conversion changes signedness: 'int' to 'char' [-Wsign-conversion]
        fillname[0]=GetDrive()+'A';
                   ~~~~~~~~~~~^~~~
dos_classes.cpp:534:17: warning: implicit conversion loses integer precision: 'unsigned long' to 'PhysPt' (aka 'unsigned int') [-Wshorten-64-to-32]
                mem_writeb(pt + i, 0x00);
```

**Comment by dreamer** [comment]
We can't merge this PR yet - I found a regression (when simply mounting a filesystem). Not sure which commit is at fault yet.

**Comment by dreamer** [comment]
Another regression (search finds only directories, not files) - introduced with 7f8a79d :/

File: src/dos/dos_classes.cpp (hunk 3 above)
Comment (dreamer) [reply src/dos/dos_classes.cpp:103]: Hmm, I don't think so. These private packed structures (like `sDIB`) are never actually used - they are only "describing" memory layout, nothing more. So if we created a constexpr object to initialize… we would still need to keep the same litany of `SSET_*` macros to actually correctly fill the emulated memory.

**Comment by dreamer** [comment]
OK, I corrected all regressions I noticed during testing; I think it's ready for another round of review :)

**Comment by dreamer** [comment]
Seems like not only PVS has problem today, but Coverity as well? https://github.com/dosbox-staging/dosbox-staging/runs/1037155189

Honestly, I don't remember how we fixed this problem last time it happened.

**Comment by kcgen** [comment]
My fault; I accidentally purged this regaining space in my google drive :/.  Will repost it shortly. Thanks for letting me know @dreamer

**Comment by dreamer** [comment]
I think this Coverity issue was actually caused by `gdown` tool. I uploaded the tarball to our "fake release page" (release in dosbox-staging.github.io repo) and it seems to be working fine.

**Comment by dreamer** [comment]
Cleaned up and rebased, let's get this over with.

@kcgen - thank you for the review :)

**Comment by kcgen** [comment]
> I think this Coverity issue was actually caused by gdown tool. I uploaded the tarball to our "fake release page" (release in dosbox-staging.github.io repo) and it seems to be working fine.

If you downloaded it, then the Google Drive must have been good. Darn, maybe Google's made another API-breaking change. Awesome that we can host it on `dosbox-staging.github.io repo`!

**Comment by dreamer** [comment]
> Awesome that we can host it on `dosbox-staging.github.io repo`!

Please make sure you can edit this page and upload new files: https://github.com/dosbox-staging/dosbox-staging.github.io/releases/tag/artifacts - in the past I used it to share 0.76.x builds with people without GitHub access, but I see no harm in adding a build tool in there (as long as licensing, etc, allows).

File: include/dos_inc.h [hunk 6]
```diff
@@ -442,30 +421,37 @@ class DOS_ParamBlock:public MemStruct {
...
+		return RealMake(seg, offsetof(sDIB, firstDPB));
+	}
+
+	void SetDeviceChainStart(uint32_t chain) { SSET_DWORD(sDIB, nulNextDriver, chain); }
+	uint32_t GetDeviceChain() const { return SGET_DWORD(sDIB, nulNextDriver); }
+
+	void SetUMBChainState(uint8_t state) { SSET_BYTE(sDIB, chainingUMB, state); }
+	uint8_t GetUMBChainState() const { return SGET_BYTE(sDIB, chainingUMB); }
+
+	void SetStartOfUMBChain(uint16_t seg) { SSET_WORD(sDIB, startOfUMBChain, seg); }
```
Comment (kcgen) [inline include/dos_inc.h:442]: @dreamer - belated comment; just caught that the `seg` argument hides the `seg` member variable.
What can we rename the argument to?

**Pull request merged**